iroh-net/src/magicsock.rs

```diff
@@ -2032,11 +1990,15 @@ impl Actor {
         // The STUN address(es) are always first.
         // Despite this sorting, clients are not relying on this sorting for decisions;
 
-        if self.inner.endpoints.write().set(&eps) {
-            log_endpoint_change(&eps);
-            if let Some(ref cb) = self.inner.on_endpoints {
-                cb(&eps[..]);
-            }
+        let updated = self
+            .inner
+            .endpoints
+            .update(DiscoveredEndpoints::new(eps))
+            .is_ok();
+        if updated {
+            let eps = self.inner.endpoints.read();
```

Comment:
I think this code is fine.  But seeing `.read()` here did make me go round and check all the docs on what exactly we're reading, who we're notifying and why we must have a lock.  Instinctively I would have cloned the endpoints before storing and then worked of the clone.  It's much of a muchness though.

Reply:
I did it this way, to avoid having to clone in all cases, and only if we actually need to do so